shared/src/main/scala/coop/rchain/store/LazyAdHocKeyValueCache.scala

```diff
@@ -0,0 +1,60 @@
+package coop.rchain.store
+
+import cats.Applicative
+import cats.effect.Concurrent
+import cats.effect.concurrent.{Deferred, Ref}
+import cats.syntax.all._
+
+trait KeyValueCache[F[_], K, V] {
+  def get(key: K, fallback: F[V]): F[V]
+  def put(key: K, value: V): F[Unit]
+  def toMap: F[Map[K, V]]
+}
+
+final case class NoOpKeyValueCache[F[_]: Applicative, K, V]() extends KeyValueCache[F, K, V] {
+  override def get(key: K, fallback: F[V]): F[V] = fallback
+  override def put(key: K, value: V): F[Unit]    = ().pure[F]
+  override def toMap: F[Map[K, V]]               = Map.empty[K, V].pure[F]
+}
+
+class LazyAdHocKeyValueCache[F[_]: Concurrent, K, V] private[LazyAdHocKeyValueCache] (
```

Comment:
I think `LazyKeyValueCache` and `LazyAdHocKeyValueCache` are totally duplicates. Do I miss something?

Reply:
One of them requires function to populate cache value during object construction, another - when calling get method. I agree it is better to have one API, but I'm not sure whats the best way do to that.

Reply:
@zsluedem Let's move refactoring of `KeyValueCache` to another PR. I created an issue #3340 to refactor get method. You can add additional stuff to that issue.

Reply:
I am ok to refactor this in another pr. I thought I miss something.